### PR TITLE
Feat/msdk 571 history of agents

### PR DESCRIPTION
**Pull request by shuffledex**

### CODE REVIEW

File: src/api/entities/SecurityToken/index.ts [hunk 1]
```diff
@@ -457,6 +459,34 @@ export class SecurityToken extends Entity<UniqueIdentifiers, string> {
    */
   public controllerTransfer: ProcedureMethod<ControllerTransferParams, void>;
 
+  /**
+   * Retrieve all Agent Operation History
```
Comment (monitz87) [inline src/api/entities/SecurityToken/index.ts:463]:
```suggestion
   * Retrieve this Security Token's Operation History
   *
   * @note Operations are grouped by the Agent Identity who performed them
```

File: src/api/entities/__tests__/Account.ts [hunk 2]
```diff
@@ -164,6 +164,7 @@ describe('Account class', () => {
             block_id: blockNumber1.toNumber(),
             address: address,
```
Comment (monitz87) [inline src/api/entities/__tests__/Account.ts:165]:
```suggestion
            address,
```

File: src/middleware/queries.ts [hunk 3]
```diff
@@ -681,3 +682,30 @@ export function getHistoryOfPaymentEventsForCa(
     variables,
   };
 }
+
+/**
+ * @hidden
+ *
+ * Get the history of each external agent in ticker
```
Comment (monitz87) [inline src/middleware/queries.ts:689]:
```suggestion
 * Get the transaction history of each external agent of a token
```

File: src/types/index.ts [hunk 4]
```diff
@@ -1113,6 +1113,11 @@ export interface TokenWithGroup {
   group: KnownPermissionGroup | CustomPermissionGroup;
 }
 
+export interface HistoricAgentOperation {
```
Comment (monitz87) [inline src/types/index.ts:1116]:
```suggestion
/**
 * Events triggered by transactions performed by an Agent Identity, related to the Token's configuration
 *   For example: changing compliance requirements, inviting/removing agent Identities, freezing/unfreezing transfers
 *
 * Token transfers (settlements or movements between Portfolios) do not count as Operations
 */
export interface HistoricAgentOperation {
```

File: src/types/index.ts [hunk 5]
```diff
@@ -1113,6 +1113,11 @@ export interface TokenWithGroup {
...
+  identity: Identity;
```
Comment (monitz87) [inline src/types/index.ts:1117]:
```suggestion
  /**
   * Agent Identity that performed the operations
   */
  identity: Identity;
```

File: src/types/index.ts [hunk 6]
```diff
@@ -1113,6 +1113,11 @@ export interface TokenWithGroup {
...
+  history: EventIdentifier[];
```
Comment (monitz87) [inline src/types/index.ts:1118]:
```suggestion
  /**
   * list of Token Operation Events that were triggered by the Agent Identity
   */
  history: EventIdentifier[];
```

**Review by monitz87 — CHANGES_REQUESTED** [review]

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/api/entities/SecurityToken/index.ts [hunk 7]
```diff
@@ -457,6 +459,36 @@ export class SecurityToken extends Entity<UniqueIdentifiers, string> {
...
   public controllerTransfer: ProcedureMethod<ControllerTransferParams, void>;
 
+  /**
+   * Retrieve this Security Token's Operation History
+   *
+   * @note Operations are grouped by the Agent Identity who performed them
+   *
+   * @note uses the middleware
+   */
+  public async getOperationHistory(): Promise<HistoricAgentOperation[]> {
```
[the remark this answers is not in the document]
Comment (shuffledex) [reply src/api/entities/SecurityToken/index.ts:469]: No. The middleware endpoint can't be paginated as it does stream processing.

**Review by monitz87 — APPROVED** [review]

**Pull request merged**